src/Init/Data/String/Basic.lean

```diff
@@ -290,73 +290,111 @@ where go (acc : String) (s : String) : List String → String
   | a :: as => go (acc ++ s ++ a) s as
   | []      => acc
 
-/-- Iterator for `String`. That is, a `String` and a position in that string. -/
+/-- Iterator over the characters (`Char`) of a `String`.
+
+Typically created by `s.iter`, where `s` is a `String`. -/
 structure Iterator where
+  /-- The string the iterator is for. -/
   s : String
+  /-- The current position.
+
+  This position is not necessarily legal for the string, for instance if one keeps calling
+  `Iterator.next` when `Iterator.atEnd` is true. If the position is not a legal one, then the
+  current character is `(default : Char)`, similar to `String.get` on an illegal position. -/
   i : Pos
   deriving DecidableEq
 
+/-- The iterator positioned at the first character of a string (if any). -/
 def mkIterator (s : String) : Iterator :=
   ⟨s, 0⟩
 
+@[inherit_doc mkIterator]
 abbrev iter := mkIterator
 
+/-- The notion of size for a string iterator is the number of bytes remaining. -/
 instance : SizeOf String.Iterator where
   sizeOf i := i.1.utf8ByteSize - i.2.byteIdx
 
 theorem Iterator.sizeOf_eq (i : String.Iterator) : sizeOf i = i.1.utf8ByteSize - i.2.byteIdx :=
   rfl
 
 namespace Iterator
+/-- The *full* string the iterator is for. -/
 def toString : Iterator → String
   | ⟨s, _⟩ => s
 
+/-- Number of bytes remaining in the iterator. -/
 def remainingBytes : Iterator → Nat
   | ⟨s, i⟩ => s.endPos.byteIdx - i.byteIdx
 
+/-- Current character position. -/
 def pos : Iterator → Pos
   | ⟨_, i⟩ => i
 
+/-- The character at the current position.
+
+On an illegal position, returns `(default : Char)`. -/
```

Comment:
```suggestion
On an invaid position, returns `(default : Char)`. -/
```

Or even
```suggestion
If the iterator points outside the bounds of the string, this returns `(default : Char)`. -/
```

Reply:
Being outside of the bounds of the string is not the only way to have an invalid position if there are UTF8 characters.

```lean
def main : IO Unit := do
  let i : String.Iterator := ⟨"🙀", ⟨1⟩⟩
  println! "curr: `{i.curr}`"
  -- curr: `A`
```

This should not happen on an iterator that's used the way it's supposed to be, but there's no proof that the byte index actually corresponds to a character so...